File: ios/Sources/ACTemplating/ExpressionCache.swift
import Foundation

/// Thread-safe cache for parsed expression ASTs
/// Ported from production Teams-AdaptiveCards-Mobile SDK's FunctionCallCache pattern
///
/// Caches parsed `Expression` ASTs to avoid re-parsing the same expression string.
/// Uses LRU eviction and optional TTL for entries.
public final class ExpressionCache: @unchecked Sendable {
    /// Cache entry wrapping a parsed expression with metadata
    private struct CacheEntry {
        let expression: Expression
        let createdAt: Date
        var lastAccessedAt: Date
        var accessCount: Int
    }

    private var cache: [String: CacheEntry] = [:]
    private let lock = NSLock()
    private let maxEntries: Int
    private let ttl: TimeInterval?

    /// Cache statistics
    public private(set) var hits: Int = 0
    public private(set) var misses: Int = 0

    /// Hit rate as a percentage (0.0 - 1.0)
    public var hitRate: Double {
        let total = hits + misses
        return total > 0 ? Double(hits) / Double(total) : 0
    }

    /// Number of entries currently in the cache
    public var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return cache.count
    }

    /// Create a new expression cache
    /// - Parameters:
    ///   - maxEntries: Maximum number of cached expressions (default: 256)
    ///   - ttl: Time-to-live for cache entries in seconds (nil = no expiration)
    public init(maxEntries: Int = 256, ttl: TimeInterval? = nil) {
        self.maxEntries = maxEntries
        self.ttl = ttl
    }

    /// Get a cached expression, or parse and cache it
    /// - Parameters:
    ///   - expression: The expression string
    ///   - parser: The parser to use if the expression is not cached
    /// - Returns: The parsed expression
    public func getOrParse(_ expressionString: String, using parser: ExpressionParser) throws -> Expression {
        lock.lock()

        // Check cache
        if var entry = cache[expressionString] {
            // Check TTL
            if let ttl = ttl, Date().timeIntervalSince(entry.createdAt) > ttl {
                cache.removeValue(forKey: expressionString)
                lock.unlock()
                // Fall through to parse
            } else {
                entry.lastAccessedAt = Date()
                entry.accessCount += 1
                cache[expressionString] = entry
                hits += 1
                lock.unlock()
                return entry.expression
            }
        } else {
            lock.unlock()
        }

        // Parse
        let parsed = try parser.parse(expressionString)

        // Store in cache
        lock.lock()
        defer { lock.unlock() }

        // Evict if at capacity
        if cache.count >= maxEntries {
            evictLRU()
        }

        misses += 1
        cache[expressionString] = CacheEntry(
            expression: parsed,
            createdAt: Date(),
            lastAccessedAt: Date(),
            accessCount: 1
        )

        return parsed
    }

    /// Clear all cached expressions
    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        cache.removeAll()
        hits = 0
        misses = 0
    }

    /// Remove expired entries (only relevant when TTL is set)
    public func purgeExpired() {
        guard let ttl = ttl else { return }

        lock.lock()
        defer { lock.unlock() }

        let now = Date()
        cache = cache.filter { _, entry in
            now.timeIntervalSince(entry.createdAt) <= ttl
        }
    }

    // MARK: - Private

    /// Evict the least recently used entry
    private func evictLRU() {
        guard let lruKey = cache.min(by: { $0.value.lastAccessedAt < $1.value.lastAccessedAt })?.key else {
            return
        }
        cache.removeValue(forKey: lruKey)
    }
}

File: android/ac-templating/src/main/kotlin/com/microsoft/adaptivecards/templating/ExpressionCache.kt
package com.microsoft.adaptivecards.templating

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

/**
 * Thread-safe cache for parsed expression ASTs.
 * Ported from production Teams-AdaptiveCards-Mobile SDK's FunctionCallCache pattern.
 *
 * Caches parsed [Expression] ASTs to avoid re-parsing the same expression string.
 * Uses LRU eviction and optional TTL for entries.
 *
 * @param maxEntries Maximum number of cached expressions (default: 256)
 * @param ttlMillis Time-to-live for cache entries in milliseconds (null = no expiration)
 */
class ExpressionCache(
    private val maxEntries: Int = 256,
    private val ttlMillis: Long? = null
) {
    private data class CacheEntry(
        val expression: Expression,
        val createdAt: Long = System.currentTimeMillis(),
        var lastAccessedAt: Long = System.currentTimeMillis(),
        var accessCount: Int = 1
    )

    private val cache = ConcurrentHashMap<String, CacheEntry>()
    private val _hits = AtomicInteger(0)
    private val _misses = AtomicInteger(0)

    /** Number of cache hits */
    val hits: Int get() = _hits.get()

    /** Number of cache misses */
    val misses: Int get() = _misses.get()

    /** Hit rate as a ratio (0.0 - 1.0) */
    val hitRate: Double
        get() {
            val total = hits + misses
            return if (total > 0) hits.toDouble() / total else 0.0
        }

    /** Number of entries currently in the cache */
    val count: Int get() = cache.size

    /**
     * Get a cached expression, or parse and cache it.
     *
     * @param expressionString The expression string
     * @param parser The parser to use if the expression is not cached
     * @return The parsed expression
     */
    fun getOrParse(expressionString: String, parser: ExpressionParser): Expression {
        // Check cache
        cache[expressionString]?.let { entry ->
            // Check TTL
            if (ttlMillis != null && System.currentTimeMillis() - entry.createdAt > ttlMillis) {
                cache.remove(expressionString)
            } else {
                entry.lastAccessedAt = System.currentTimeMillis()
                entry.accessCount++
                _hits.incrementAndGet()
                return entry.expression
            }
        }

        // Parse
        val parsed = parser.parse(expressionString)

        // Evict if at capacity
        if (cache.size >= maxEntries) {
            evictLRU()
        }

        _misses.incrementAndGet()
        cache[expressionString] = CacheEntry(expression = parsed)

        return parsed
    }

    /** Clear all cached expressions */
    fun clear() {
        cache.clear()
        _hits.set(0)
        _misses.set(0)
    }

    /** Remove expired entries (only relevant when TTL is set) */
    fun purgeExpired() {
        val ttl = ttlMillis ?: return
        val now = System.currentTimeMillis()
        cache.entries.removeIf { now - it.value.createdAt > ttl }
    }

    private fun evictLRU() {
        cache.entries
            .minByOrNull { it.value.lastAccessedAt }
            ?.let { cache.remove(it.key) }
    }
}
